Reduce.go

```diff
@@ -0,0 +1,326 @@
+			return nil, err
+		}
+		return newLiteral(value), nil
+	}
+
+	// optimizations
+	switch stage.symbol {
+	case AND:
+		if leftIsLiteral && left == true {
+			// true && x -> x
+			return rightStage, nil
+		}
+		if rightIsLiteral && right == true {
+			// x && true -> x
+			return leftStage, nil
+		}
+		if rightIsLiteral && right == false {
+			// x && false -> false
+			return newLiteral(false), nil
+		}
+
+	case OR:
+		if leftIsLiteral && left == false {
+			// false || x -> x
+			return rightStage, nil
+		}
+		if rightIsLiteral && right == false {
+			// x || false -> x
+			return leftStage, nil
+		}
+		if rightIsLiteral && right == true {
+			// x || true -> true
+			return newLiteral(true), nil
+		}
+
+	case PLUS:
+		if leftIsLiteral && left == 0.0 {
+			// 0 + x -> x
+			return rightStage, nil
+		}
+		if rightIsLiteral && right == 0.0 {
+			// x + 0 -> x
+			return leftStage, nil
+		}
+
+	case MINUS:
+		if leftIsLiteral && left == 0.0 {
+			// 0 - x -> -x
+			return &evaluationStage{
+				symbol:     NEGATE,
+				rightStage: rightStage,
+				operator:   negateStage,
+			}, nil
+		}
+		if rightIsLiteral && right == 0.0 {
+			// x - 0 -> x
+			return leftStage, nil
+		}
+
+	case MULTIPLY:
+		if leftIsLiteral && left == 0.0 || rightIsLiteral && right == 0.0 {
+			// 0 * x -> 0, x * 0 -> 0
+			return newLiteral(0.0), nil
+		}
+		if leftIsLiteral && left == 1.0 {
+			// 1 * x -> x
+			return rightStage, nil
+		}
+		if rightIsLiteral && right == 1.0 {
+			// x * 1 -> x
+			return leftStage, nil
+		}
+
+	case DIVIDE:
+		if rightIsLiteral && right == 1.0 {
+			// x / 1 -> x
+			return leftStage, nil
+		}
+
+	case COALESCE:
+		if rightIsLiteral && right == nil {
+			// x ?: nil -> x
+			return leftStage, nil
+		}
+
+	case TERNARY_TRUE:
+		if leftIsLiteral && left == true {
+			// true ? x : y -> x
+			return rightStage, nil
+		}
+
+	case TERNARY_FALSE:
+		if leftStage == nil {
+			// false ? x : y -> y
+			return rightStage, nil
+		}
+		if leftStage.symbol != TERNARY_TRUE {
+			// true ? x : y -> x
+			return leftStage, nil
+		}
+	}
+
+	return &evaluationStage{
+		symbol:          stage.symbol,
+		leftStage:       leftStage,
+		rightStage:      rightStage,
+		operator:        stage.operator,
+		leftTypeCheck:   stage.leftTypeCheck,
+		rightTypeCheck:  stage.rightTypeCheck,
+		typeCheck:       stage.typeCheck,
+		typeErrorFormat: stage.typeErrorFormat,
+	}, nil
+}
+
+func newLiteral(value interface{}) *evaluationStage {
+	return &evaluationStage{
+		symbol:   LITERAL,
+		operator: makeLiteralStage(value),
+	}
+}
+
+func getLiteralValue(stage *evaluationStage) (interface{}, bool) {
+	if stage != nil && stage.symbol == LITERAL {
+		value, err := stage.operator(nil, nil, nil)
+		if err == nil {
+			return value, true
+		}
+	}
+	return nil, false
+}
+
+func buildTokens(stage *evaluationStage) ([]ExpressionToken, error) {
+	if stage == nil {
+		return []ExpressionToken{}, nil
+	}
+
+	switch stage.symbol {
+	case VALUE:
+		p := testParameters{}
```

Comment:
I'm a little unclear on what's happening here. Can you explain a little about this case?

Reply:
It looks like the only way to get variable name without extending `evaluationStage` is to actually invoke the operator and look what variable it queried from parameters, because the name is captured in an operator closure. `testParameters` (I know it's a bad name, but didn't think of a better one yet) is a special `Parameters` implementation that remembers the last parameter name it was asked.

Reply:
Naming is hard. Maybe something like 'capturedParams' is better?
Makes sense. You can probably throw in that same explanation as a comment.